zerostash/src/commands/mount.rs

```diff
@@ -0,0 +1,279 @@
+    /// Start the application.
+    async fn run(&self) {
+        let stash = self.stash.open();
+
+        if let Err(e) = mount(stash, &self.options, &self.mount_point) {
+            panic!("Error = {}", e)
+        }
+    }
+}
+
+fn mount(
+    stash: Infinitree<Files>,
+    options: &restore::Options,
+    mountpoint: &str,
+) -> anyhow::Result<()> {
+    let (tx, finished) = mpsc::sync_channel(2);
+    let destroy_tx = tx.clone();
+    ctrlc::set_handler(move || tx.send(()).expect("Could not send signal on channel."))
+        .expect("Error setting Ctrl-C handler");
+
+    let filesystem = SimpleFs::open(stash, options, destroy_tx).unwrap();
+    let fuse_args = [OsStr::new("-o"), OsStr::new("fsname=zerostash")];
+
+    let fs = fuse_mt::FuseMT::new(filesystem, 1);
+
+    // Mount the filesystem.
+    let handle = spawn_mount(fs, mountpoint, &fuse_args[..])?;
+
+    // Wait until we are done.
+    finished.recv().expect("Could not receive from channel.");
+
+    // Ensure the filesystem is unmounted.
+    handle.join();
+
+    Ok(())
+}
+
+type DirPath = PathBuf;
+type FilePath = PathBuf;
+type FileContent = Vec<u8>;
+type Metadata = Arc<Entry>;
+
+struct SimpleFs {
+    destroy_tx: mpsc::SyncSender<()>,
+    stash: Infinitree<Files>,
+    file_map: HashMap<FilePath, Metadata>,
+    file_content: Mutex<HashMap<FilePath, FileContent>>,
+    dir_map: HashMap<DirPath, Vec<DirectoryEntry>>,
+    file_attr_memory: Mutex<HashMap<PathBuf, FileAttr>>,
+}
+
+fn add_dir_to_map(
+    dir_map: &mut HashMap<DirPath, Vec<DirectoryEntry>>,
+    path: &Path,
+    kind: FileType,
+) {
+    let name = path
+        .file_name()
+        .expect("All files have filenames")
+        .to_owned();
+
+    let parent = path
+        .parent()
+        .expect("Paths should have parents")
+        .to_path_buf();
+
+    if !dir_map.contains_key(&parent) {
+        add_dir_to_map(dir_map, &parent, FileType::Directory);
+    }
+
+    dir_map
+        .entry(parent)
+        .or_default()
+        .push(DirectoryEntry { name, kind });
+}
+
+const DIR_ATTR: FileAttr = FileAttr {
+    size: 0,
+    blocks: 0,
+    atime: SystemTime::UNIX_EPOCH,
+    mtime: SystemTime::UNIX_EPOCH,
+    ctime: SystemTime::UNIX_EPOCH,
+    crtime: SystemTime::UNIX_EPOCH,
+    kind: FileType::Directory,
+    perm: 0o444,
+    nlink: 1,
+    uid: 1000,
+    gid: 1000,
+    rdev: 0,
+    flags: 0,
+};
+
+pub fn match_filetype(file_type: &zerostash_files::FileType) -> FileType {
+    match file_type {
+        zerostash_files::FileType::File => FileType::RegularFile,
+        zerostash_files::FileType::Symlink(_) => FileType::Symlink,
+        zerostash_files::FileType::Directory => panic!("Didnt expect a directory!"),
+    }
+}
+
+impl SimpleFs {
+    pub fn open(
+        stash: Infinitree<Files>,
+        options: &restore::Options,
+        destroy_tx: mpsc::SyncSender<()>,
+    ) -> Result<Self> {
+        let mut dir_map: HashMap<DirPath, Vec<DirectoryEntry>> = HashMap::new();
```

Comment:
Do you need to specify the type here? I think based on below it should be inferred.